### PR TITLE
Update Job Deletion Logic To Support Force Kill

### PR DESCRIPTION
**Pull request by d80tb7**
This changes the executor's job deletion logic so that:

- The first attempt to delete a job attempts to delete the job while respecting the graceful termination period set on the pod
- Any subsequent delete attempts that occur within the graceful termination period + killTimeout (killTimeout defaults to 5 minutes but is configurable) are no-ops.
- Any subsequent delete attempts that occur after graceful termination period + killTimeout attempts to delete the job with a graceful termination of 0 (i.e. force kills the job)

### CODE REVIEW

File: internal/executor/context/cluster_context_test.go [hunk 1]
```diff
@@ -308,6 +313,33 @@ func TestKubernetesClusterContext_ProcessPodsToDelete_CallPatchOnClient_WhenPods
...
+	pod = createSubmittedBatchPod(t, clusterContext)
+	pod.DeletionGracePeriodSeconds = pointer.Int64(10)
+	pod.DeletionTimestamp = &metav1.Time{Time: testClock.Now().Add(-310 * time.Second)}
+	client.Fake.ClearActions()
+
+	// submitPodsWithWait(t, clusterContext, pod)
+	client.Fake.ClearActions()
+	clusterContext.DeletePods([]*v1.Pod{pod})
+	clusterContext.ProcessPodsToDelete()
+	assert.Equal(t, len(client.Fake.Actions()), 2)
```
Comment (d80tb7) [inline internal/executor/context/cluster_context_test.go:340]: Here I want to assert that we using a timout of 0 to kill the pod, but it doesn't look like the fake client tells me this.

File: internal/executor/context/cluster_context.go [hunk 2]
```diff
@@ -334,36 +343,55 @@ func (c *KubernetesClusterContext) DeleteIngress(ingress *networking.Ingress) er
...
+			// We've never tried to delete this pod before.  Delete using the grace period
+			c.doDelete(podToDelete, false)
+		} else {
+			// we've tried to delete this pod before. If we're after the kill period then force delete
+			// else it's a no-op
+			killTime := podToDelete.DeletionTimestamp.
+				Add(util.GetDeletionGracePeriodOrDefault(podToDelete)).
+				Add(c.killTimeout)
+			if c.clock.Now().After(killTime) {
+				log.Warnf("Pod %s/%s was requested deleted at %s, but is still present.  Force killing.", podToDelete.Namespace, podToDelete.Name, podToDelete.DeletionTimestamp)
```
Comment (samclark) [inline internal/executor/context/cluster_context.go:360]: shall we make this INFO level?

File: internal/executor/context/cluster_context.go [hunk 3]
```diff
@@ -82,6 +86,8 @@ type KubernetesClusterContext struct {
 	eventInformer            informer.EventInformer
 	// If provided, stops object creation while EtcdMaxFractionOfStorageInUse or more of etcd storage is full.
 	etcdHealthMonitor healthmonitor.EtcdLimitHealthMonitor
+	killTimeout       time.Duration
```
Comment (samclark) [inline internal/executor/context/cluster_context.go:89]: s/killTimeout/podKillTimeout

**Review by samclark — DISMISSED** [review]

File: internal/executor/context/cluster_context.go (hunk 3 above)
Comment (d80tb7) [reply internal/executor/context/cluster_context.go]: done

File: internal/executor/context/cluster_context.go (hunk 2 above)
Comment (d80tb7) [reply internal/executor/context/cluster_context.go]: done

**Review by samclark — APPROVED** [review]

**Pull request merged**